### PR TITLE
Bluetooth: host: Fix non-extended adverts not being limited to 31 bytes

### LINKED ISSUE #38745: Bluetooth when configured for extended advertising does not limit advertisement packet size if a non-extended avertisement is used
**Describe the bug**
Using nRF connect SDK 1.6.1 based on zephyr 2.6 with an nRF52840, in subsys/bluetooth/host/adv.c it seems that when advertisements and scan reports are generated when the code has been build with extended advertising support but starting using NON-extended advertising, the maximum packet sizes are incorrectly set to the extended advertising lengths, so if you have a device name of length 50 characters, it will put all 50 characters into the packet and then complain that the packet is too large even though the shortened name should be used instead. In the `hci_set_ad_ext` function is:
```
        err = set_data_add(set_data->data, BT_HCI_LE_EXT_ADV_FRAG_MAX_LEN,
                           ad, ad_len, &set_data->len);
```
But the second parameter is valid only for extended advertisements, for non-extended advertisements the hard limit is 31.

**To Reproduce**
Create an application with support for extended advertising and set a large (e.g. 50 character) device name and advertise in legacy mode

**Expected behavior**
For the device name to be limited and maximum packet size to be 31 bytes

**Impact**
Showstopper, legacy advertising is broken

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain: Zephyr SDK 0.13.1
 - Commit SHA or Version used: ncs 1.6.1 using zephyr 2.6

### PR DESCRIPTION
**Pull request by nordicjm**
Fixes an issue whereby the application is configured for extended
advertising mode but advertises in legacy mode with a large device name
which should be limited to 31 bytes

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/38745

Signed-off-by: Jamie McCrae <jamie.mccrae@lairdconnect.com>

Not sure if the initial checking is required or not

### CODE REVIEW

**Comment by lairdjm** [comment]
DNM this, the original comment in the post of using `atomic_test_bit(adv->flags, BT_ADV_EXT_ADV)` to check if an advert is extended or not is invalid

File: subsys/bluetooth/host/adv.c [hunk 1]
```diff
@@ -422,8 +422,13 @@ static int hci_set_adv_ext_complete(struct bt_le_ext_adv *adv, uint16_t hci_op,
 	set_data = net_buf_add(buf, sizeof(*set_data));
 	(void)memset(set_data, 0, sizeof(*set_data));
 
-	err = set_data_add_complete(set_data->data, BT_HCI_LE_EXT_ADV_FRAG_MAX_LEN,
+	err = set_data_add_complete(set_data->data,
+				    (atomic_test_bit(adv->flags,
+						     BT_ADV_EXT_ADV) ?
+					BT_HCI_LE_EXT_ADV_FRAG_MAX_LEN :
+					BT_GAP_ADV_MAX_ADV_DATA_LEN),
```
Comment (Thalley) [inline subsys/bluetooth/host/adv.c:429]: For readability sake, can you put this into a (const) variable?

**Review by Thalley — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/host/adv.c [hunk 2]
```diff
@@ -535,14 +540,16 @@ static int hci_set_ad_ext(struct bt_le_ext_adv *adv, uint16_t hci_op,
...
+	    atomic_test_bit(adv->flags, BT_ADV_EXT_ADV)) {
 		/* It is not allowed to set advertising data in multiple
 		 * operations while the advertiser is running.
 		 */
 		return -EAGAIN;
 	}
 
-	if (total_len_bytes <= BT_HCI_LE_EXT_ADV_FRAG_MAX_LEN) {
+	if (total_len_bytes <= BT_HCI_LE_EXT_ADV_FRAG_MAX_LEN ||
+	    !atomic_test_bit(adv->flags, BT_ADV_EXT_ADV)) {
```
Comment (rugeGerritsen) [inline subsys/bluetooth/host/adv.c:552]: I would suggest to remove the additional checks for advertising type when determining if complete or fragmented data will be used. That makes the behavior consistent for legacy and extended advertising when using extended advertising commands.

This approach will disallow setting data with a local name longer than `BT_HCI_LE_EXT_ADV_FRAG_MAX_LEN` while the advertiser is running even if we could have truncated the name. We can perform a dry run to determine if a truncation of advertising name is possible, but IMO this increases the complexity too much.

Another suggestion is to disallow truncating of advertising name when using extended commands and extended advertising type. That would require us to put an additional check inside `hci_set_adv_ext_complete()`

**Review by Thalley — APPROVED** [review]

**Review by joerchan — APPROVED** [review]
Thank you for the fix.

**Review by alwa-nordic — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

**Pull request merged**